### PR TITLE
feat: Add alter table procedure

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR adds a procedure `AlterTableProcedure` to alter a table.
- It invokes `TableEngineProcedure::alter_table_procedure()` to alter the metadata of the table
- If the alter request is renaming, it also renames the table in the catalog manager

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #286

### CODE REVIEW

File: src/datanode/src/sql/alter.rs [hunk 1]
```diff
@@ -13,18 +13,25 @@
...
 use table::engine::{EngineContext, TableReference};
 use table::requests::{AddColumnRequest, AlterKind, AlterTableRequest};
+use table_procedure::AlterTableProcedure;
 
 use crate::error::{self, Result};
 use crate::sql::SqlHandler;
 
 impl SqlHandler {
     pub(crate) async fn alter(&self, req: AlterTableRequest) -> Result<Output> {
+        if let Some(procedure_manager) = &self.procedure_manager {
```
Comment (fengjiachun) [inline src/datanode/src/sql/alter.rs:31]: Is this the final form（code style） in which `procedure` is used?

```
if has procedure {
    do something with procedure
} else {
    do the same thing without procedure
}
```

Comment (evenyag) [reply src/datanode/src/sql/alter.rs:31]: Maybe we can always use procedures to do this when procedures for all table operations are ready. Then we don't need an `if-else` check.

File: src/table-procedure/src/error.rs [hunk 2]
```diff
@@ -51,11 +51,17 @@ pub enum Error {
     #[snafu(display("Schema {} not found", name))]
     SchemaNotFound { name: String },
 
+    #[snafu(display("Table {} not found", name))]
+    TableNotFound { name: String },
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/table-procedure/src/error.rs:55]: table not found has different status code and message, so I add a new error variant.

File: src/datanode/src/sql/alter.rs (hunk 1 above)
Comment (fengjiachun) [reply src/datanode/src/sql/alter.rs:31]: A TODO here?

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/table-procedure/src/alter.rs [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+            let rename_req = RenameTableRequest {
+                catalog: self.data.request.catalog_name.clone(),
+                schema: self.data.request.schema_name.clone(),
+                table_name: self.data.request.table_name.clone(),
+                new_table_name: new_table_name.clone(),
+                table_id,
+            };
+
+            self.catalog_manager
+                .rename_table(rename_req)
```
Comment (v0y4g3r) [inline src/table-procedure/src/alter.rs:248]: Just find out that if to procedure runs concurrently to rename two different table to the same name, the later rename operation will override previous table with the same name. 

https://github.com/GrepTimeTeam/greptimedb/blob/da66138e809975e1f88b80cd8cb0457de1946e8a/src/catalog/src/local/memory.rs#L325-L337

File: src/datanode/src/sql/alter.rs (hunk 1 above)
Comment (evenyag) [reply src/datanode/src/sql/alter.rs:31]: There is a related todo in SqlHandler, we need to support multiple table engines and then we can enable the procedure framework by default.

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**